### PR TITLE
Support for CancellationToken in SiloHost, AzureSilo & WindowsServerHost

### PR DESCRIPTION
**Pull request by lnaie**
A draft that doesn't impact existing functionality. Extended a few methods with an extra CancellationToken param.

### CODE REVIEW

**Comment by gabikliot** [comment]
Great, thank you, we will look into it.

File: src/OrleansAzureUtils/AzureSilo.cs [hunk 1]
```diff
@@ -227,11 +228,29 @@ public void Run()
...
                 throw new ApplicationException("Silo failed to start correctly - aborting");
         }
 
+		/// <summary>
+		/// Makes this Orleans silo begin executing and become active.
+		/// Note: This method call will only return control back to the caller when the silo is shutdown or 
+		/// an external request for cancellation has been issued.
+		/// </summary>
+		/// <param name="cancellationToken">Cancellation token.</param>
+		public void Run(CancellationToken cancellationToken)
```
Comment (gabikliot) [inline src/OrleansAzureUtils/AzureSilo.cs]: Can we eliminate the code duplication, by having Run and Run(CancellationToken) both call Run_Impl(CancellationToken) and check CancellationToken  for null inside.

File: src/OrleansHost/WindowsServerHost.cs [hunk 2]
```diff
@@ -91,6 +92,49 @@ public int Run()
...
+		/// <param name="cancellationToken">Cancellation token.</param>
+		/// <remarks>
+		/// If the Silo starts up successfully, then this method will block and not return 
+		/// until the silo shutdown event is triggered or the silo shuts down for some other reason or 
+		/// an external request for cancellation has been issued.
+		/// If the silo fails to star, then a StartupError.txt summary file will be written, 
+		/// and a process mini-dump will be created in the current working directory.
+		/// </remarks>
+		/// <returns>Returns <c>false</c> is Silo failed to start up correctly.</returns>
+		public int Run(CancellationToken cancellationToken)
```
Comment (gabikliot) [inline src/OrleansHost/WindowsServerHost.cs]: Same question here: can we combine both Runs into 1 implementation?

File: src/OrleansRuntime/Silo/SiloHost.cs [hunk 3]
```diff
@@ -250,6 +250,36 @@ public void WaitForOrleansSiloShutdown()
...
+		/// <summary>
+		/// Wait for this silo to shutdown or be cancelled.
+		/// </summary>
+		/// <param name="cancellationToken">Cancellation token.</param>
+		/// <remarks>
+		/// Note: This method call will block execution of current thread, 
+		/// and will not return control back to the caller until the silo is shutdown or 
+		/// an external request for cancellation has been issued.
+		/// </remarks>
+		public void WaitForOrleansSiloShutdown(CancellationToken cancellationToken)
```
Comment (gabikliot) [inline src/OrleansRuntime/Silo/SiloHost.cs]: And here we can just rename the one WaitForOrleansSiloShutdown into WaitForOrleansSiloShutdown(CancellationToken cancellationToken=null) and check for null  inside and thus avoid the code dup.

**Comment by gabikliot** [comment]
Looks good, except for some minor comments.

**Comment by lnaie** [comment]
Fair points, but I had to change the signature of the methods.

**Comment by jthelin** [comment]
LGTM. 

BTW, I just wanted to call out one best-practice topic for public APIs which @lnaie did great here in this change, but which is sometimes quite counter-intuitive to many people [including some members of the core Orleans Team :wink: ]. 

``` c#
public int Run() { ... }
public int Run(CancellationToken cancellationToken) { ... }
```

Default parameter values in C# only have __compile-time__ effect, so adding a new [optional] parameter with a default value to an existing __public__ method actually _breaks_ binary compatibility for the assembly and so prevents drop-in replace / upgrade of new framework assemblies at run-time.

This is one of the reasons why overloading is _always_ better than adding optional parameters for changing __public__ APIs and is something specifically called out in the .NET Framework Design:

"__DO__ use member overloading rather than defining members with default arguments."
https://msdn.microsoft.com/en-us/library/ms229029(v=vs.110).aspx

**Comment by gabikliot** [comment]
That is what I suggested: 2 public overloads and both use the same private implementation. Both for Run and Wait.

**Comment by lnaie** [comment]
Added the last bits. Thanks.

**Comment by jthelin** [comment]
Cool. Merged. Thanks for the contribution @lnaie  :+1:

**Pull request merged**